### PR TITLE
Shell: Bluetooth: Add shell_get_ctx() instead of global var

### PR DESCRIPTION
**Pull request by jori-nordic**
Use a getter for the shell context, which is initialized by the shell
backend.

Signed-off-by: Jonathan Rico <jonathan.rico@nordicsemi.no>

### CODE REVIEW

File: subsys/bluetooth/mesh/shell.c [hunk 1]
```diff
@@ -410,7 +408,6 @@ static int cmd_uuid(const struct shell *shell, size_t argc, char *argv[])
 static int cmd_init(const struct shell *sh, size_t argc, char *argv[])
 {
 
-	ctx_shell = sh;
```
[the remark this answers is not in the document]
Comment (jori-nordic) [reply subsys/bluetooth/mesh/shell.c:413]: alright, works for me. There's an unrelated warning for mesh anyways, I'll let you deal with that :p

File: subsys/shell/shell.c [hunk 2]
```diff
@@ -1371,6 +1378,8 @@ int shell_init(const struct shell *shell, const void *transport_config,
 	shell->ctx->tid = tid;
 	k_thread_name_set(tid, shell->thread_name);
 
+	shell_ctx = (const struct shell *)(shell->ctx);
```
Comment (Thalley) [inline subsys/shell/shell.c:1381]: Not sure I understand this.

We have a shell of the type `struct shell`, and then you assign the `shell->ctx` (which is a `struct shell_ctx` to a variable `shell_ctx`, which has the type `struct shell`. 

Basically you are casting the `shell->ctx` to a new type; doesn't that give all sorts of problems?

**Review by jakub-uC — CHANGES_REQUESTED** [review]
In my humble opinion, this modification does not make sense because when there are more backends active then the returned context may not be what you expect.

File: subsys/shell/shell.c (hunk 2 above)
Comment (jori-nordic) [reply subsys/shell/shell.c:1381]: my bad I misread the code, fixed now

**Comment by jori-nordic** [comment]
> In my humble opinion, this modification does not make sense because when there are more backends active then the returned context may not be what you expect.

Wait, we can have multiple backends active at once ? But then the way it was before was flawed too: if you "logged in" through let's say RTT, sent the `bt_init` and then "logged in" through UART, you'd not see the callback messages on the UART but only on RTT. What do you suggest instead ?

**Comment by jakub-uC** [comment]
@jori-nordic 
The previous implementation will not work if these commands are called at the same time from different backends. In your case, if the order of shell backend initialization is not to your liking then this code will never work properly.

**Comment by jori-nordic** [comment]
> @jori-nordic The previous implementation will not work if these commands are called at the same time from different backends. In your case, if the order of shell backend initialization is not to your liking then this code will never work properly.

Yep I got that, I was thinking of maybe having a default backend (selectable thru Kconfig) that would be used when calling shell print with `NULL`.

But after discussing it with you I now think those shell prints should rather be `LOG_DBG`s. 
The user can them filter them out (see https://github.com/zephyrproject-rtos/zephyr/pull/46762/commits/b5b7d4956f2b2d574fde64eae00ae7ea6daee095 ) in the shell backends he doesn't want "polluted" by the logs.

What's your take @Thalley and @PavelVPV ?

**Comment by Thalley** [comment]
> > @jori-nordic The previous implementation will not work if these commands are called at the same time from different backends. In your case, if the order of shell backend initialization is not to your liking then this code will never work properly.
> 
> Yep I got that, I was thinking of maybe having a default backend (selectable thru Kconfig) that would be used when calling shell print with `NULL`.
> 
> But after discussing it with you I now think those shell prints should rather be `LOG_DBG`s. The user can them filter them out (see [b5b7d49](https://github.com/zephyrproject-rtos/zephyr/commit/b5b7d4956f2b2d574fde64eae00ae7ea6daee095) ) in the shell backends he doesn't want "polluted" by the logs.
> 
> What's your take @Thalley and @PavelVPV ?

Given that the shell is typically interactive, using the logging system seems odd, but not necessarily bad. 

We could have a default shell (which would just be the first that is initialized), and then also add functions to iterate over backends and select whichever one we want as the "context shell".

**Comment by jori-nordic** [comment]
@Thalley yeah something like `shell-set-ctx` which could be called with the name of the backend or without args, setting the current one as the default.

**Comment by Thalley** [comment]
> @Thalley yeah something like `shell-set-ctx` which could be called with the name of the backend or without args, setting the current one as the default.

We may also need to consider the term `context` here. `shell_ctx` kind of makes sense when a shell command is done in some context, but if we are actually talking about different shell/shell backends, another term may be useful. 

If we do something like a `set`, it could be something as simple as `shell_set_active` and `shell_get_active` or something like that.

**Comment by jakub-uC** [comment]
Using the existing code we can add command: `shell set default` or something similar. This commad can simply set newly added pointer: `shell_ctx = shell;`
Or you can add shell api function for setting this context. Next this function could be called from the bt commands.

**Comment by PavelVPV** [comment]
> > @jori-nordic The previous implementation will not work if these commands are called at the same time from different backends. In your case, if the order of shell backend initialization is not to your liking then this code will never work properly.
> 
> Yep I got that, I was thinking of maybe having a default backend (selectable thru Kconfig) that would be used when calling shell print with `NULL`.
> 
> But after discussing it with you I now think those shell prints should rather be `LOG_DBG`s. The user can them filter them out (see [b5b7d49](https://github.com/zephyrproject-rtos/zephyr/commit/b5b7d4956f2b2d574fde64eae00ae7ea6daee095) ) in the shell backends he doesn't want "polluted" by the logs.
> 
> What's your take @Thalley and @PavelVPV ?

I would prefer all messages in mesh/shell.c issued unsolicitly (through callbacks) to be sent to shell. Pushing them to the logging system creates some problems. A user will need to configure the logging properly to see them (messages) together with shell commands (which also limits the logging configuration). There may be a need to pass the messages through some other transport. Then, a user needs somehow to collect all messages from different sources.

It is quite inconvenient to have `mesh init` function that just initializes the shell context. Until it is called, any message passed from the callback will not be printed. This makes a user to either call this command every time when device boots up or call this directly from the code at the device initialization. But since (as I understand now) there can be multiple shell backends running, there is no easy way to solve this inconvenience.

**Review by alwa-nordic — CHANGES_REQUESTED** [review]
As @jakub-uC pointed out, `shell_get_ctx` does not fit into the design of Shell, because there is no singleton shell instance.

I think the cmd functions in Bluetooth Shell are not reentrant. So it's unsafe to use multiple shell instances concurrently. To avoid having to fix that now, I propose the Bluetooth Shell picks up the first shell instance that calls any Bluetooth Shell cmd and remembers that shell as its "client" / shell_ctx. Then all Bluetooth Shell cmds shall refuse to work when called from a different shell instance. For correctness, shell_ctx must become an atomic in this case.

**Comment by jori-nordic** [comment]
> As @jakub-uC pointed out, `shell_get_ctx` does not fit into the design of Shell, because there is no singleton shell instance.
> 
> I think the cmd functions in Bluetooth Shell are not reentrant. So it's unsafe to use multiple shell instances concurrently. To avoid having to fix that now, I propose the Bluetooth Shell picks up the first shell instance that calls any Bluetooth Shell cmd and remembers that shell as its "client" / shell_ctx. Then all Bluetooth Shell cmds shall refuse to work when called from a different shell instance. For correctness, shell_ctx must become an atomic in this case.

@alwa-nordic Sounds like a lot of work, do you want to do it (in a separate PR) ? I wanted to make a quick fix, but it looks like the issue is actually bigger than it seemed. 

If others are okay with me keeping the PR the way it is + adding a `shell set active` or `shell set default` for multiple-backend users, just put a thumbsup on this comment.

**Pull request closed without merging**